loki/transformations/data_offload/offload_deepcopy.py

```diff
@@ -346,3 +356,429 @@ def transform_module(self, module, **kwargs): # pylint: disable=unused-argument
+            for region in FindNodes(ir.PragmaRegion).visit(routine.body):
+
+                # Only work on active `!$loki data` regions
+                if not is_loki_pragma(region.pragma, starts_with='data'):
+                    continue
+
+                parameters = get_pragma_parameters(region.pragma, starts_with='data')
+                driver_loops = find_driver_loops(region.body, targets)
+
+                # skip the deepcopy for variables previously marked as present/private
+                present = self.get_pragma_vars(parameters, 'present')
+                private = self.get_pragma_vars(parameters, 'private')
+
+                # temporary variables are not copied back to host and are wiped from device memory
+                temporary = self.get_pragma_vars(parameters, 'temporary')
+
+                # device_resident variables are left on device (i.e. neither copied back to host nor deleted)
+                device_resident = self.get_pragma_vars(parameters, 'device_resident')
+
+                copy, host, wipe = (), (), ()
+                present_vars = ()
+                for loop in driver_loops:
+
+                    analysis = analyses[loop]
+
+                    # update analysis with manual overrides
+                    analysis = self.update_with_manual_overrides(parameters, analysis, routine.symbol_map)
+
+                    # recursively traverse analysis and generate deepcopy
+                    _copy, _host, _wipe = self.generate_deepcopy(routine, analysis=analysis, present=present,
+                                                                 private=private, temporary=temporary,
+                                                                 device_resident=device_resident,
+                                                                 typedef_configs=typedef_configs)
+
+                    copy += _copy
+                    host += _host
+                    wipe += _wipe
+
+                    present_vars += as_tuple(v.name for v in analysis if not v in private)
+
+                # replace the `!$loki data` PragmaRegion with the generated deepcopy instructions
+                pragma_map.update(self.insert_deepcopy_instructions(region, self.mode, copy, host, wipe, present_vars))
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+    def wrap_in_loopnest(self, var, body, routine):
+        """Wrap body in loop nest corresponding to the shape of var."""
+
+        # Don't wrap an empty body
+        if not body:
+            return ()
+
+        loopbody = ()
+        loop_vars = []
+        variable_map = routine.variable_map
+
+        for dim in range(len(var.type.shape)):
+            if f'j{dim+1}' in variable_map:
+                loop_vars += [variable_map[f'j{dim+1}']]
+            else:
+                loop_vars += [sym.Variable(name=f'J{dim+1}', type=SymbolAttributes(dtype=BasicType.INTEGER),
+                                           scope=routine)]
```

Comment:
For this, you might be able to use the `get_integer_variable` utility instead? https://github.com/ecmwf-ifs/loki/blob/c9f55e53f535bca0485335569015aa8ac0c8ee29/loki/transformations/utilities.py#L542